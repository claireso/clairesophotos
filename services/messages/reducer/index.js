import { useReducer, useCallback } from 'react'
import emitter from '@services/emitter'

const INITIAL_STATE = []

const ACTION_ADD_MESSAGE = 'messages/add'
const ACTION_CLOSE_MESSAGE = 'messages/remove'
const ACTION_CLOSE_ALL = 'messages/close_all'

function reducer(state, { type, ...message }) {
  switch (type) {
    case ACTION_ADD_MESSAGE: {
      let index = state.length

      if (message.key) {
        const newIndex = state.findIndex((m) => m.key === message.key)

        if (newIndex > -1) {
          index = newIndex
        }
      }

      return [...state.slice(0, index), message, ...state.slice(index + 1)]
    }

    case ACTION_CLOSE_MESSAGE: {
      const index = message.index

      return [...state.slice(0, index), ...state.slice(index + 1)]
    }

    case ACTION_CLOSE_ALL: {
      return []
    }

    default:
      return [...state]
  }
}

export const useMessagesReducer = () => {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE)

  const displayMessage = useCallback(
    (message) => dispatch({ type: ACTION_ADD_MESSAGE, ...message }),
    [dispatch]
  )
  const closeMessage = useCallback(
    (index) => dispatch({ type: ACTION_CLOSE_MESSAGE, index }),
    [dispatch]
  )
  const closeAllMessages = useCallback(
    () => dispatch({ type: ACTION_CLOSE_ALL }),
    [dispatch]
  )

  return [state, { displayMessage, closeMessage, closeAllMessages }]
}

const displayMessage = (status) => (message) => {
  emitter.emit('feat/messages/add', {
    status,
    ...message
  })
}

export const displaySuccessMessage = displayMessage('success')
export const displayErrorMessage = displayMessage('error')
