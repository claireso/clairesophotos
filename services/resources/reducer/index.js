import React, { useReducer, useContext, useCallback } from 'react'
import PropTypes from 'prop-types'

const STATUS_IDLE = 'idle'
const STATUS_PENDING = 'pending'
const STATUS_LOADING = 'loading'
const STATUS_ERROR = 'error'
const STATUS_UPDATE_DATA_SUCCESS = 'resources/update_success'

const ACTION_UPDATE_STATUS = 'resources/update_status'
const ACTION_LOAD_RESOURCES = 'resources/get_all'
const ACTION_LOAD_RESOURCE = 'resources/get'
const ACTION_ADD_RESOURCE = 'resources/add'
const ACTION_DELETE_RESOURCE = 'resources/delete'
const ACTION_EDIT_RESOURCE = 'resources/edit'

const INITIAL_STATE = {
  status: STATUS_IDLE,
  items: [],
  pager: {}
}

export const reducer = (state, action) => {
  switch (action.type) {
    case ACTION_UPDATE_STATUS:
      return {
        ...state,
        status: action.status
      }

    case ACTION_LOAD_RESOURCES:
      return {
        ...state,
        ...action.response,
        status: STATUS_UPDATE_DATA_SUCCESS
      }

    case ACTION_LOAD_RESOURCE:
      return {
        ...state,
        single: action.response
      }

    case ACTION_ADD_RESOURCE:
      return {
        ...state,
        status: STATUS_UPDATE_DATA_SUCCESS,
        items: [action.response, ...state.items],
        pager: {
          ...state.pager,
          count: state.pager.count + 1
        }
      }

    case ACTION_EDIT_RESOURCE: {
      const index = state.items.findIndex((p) => p.id == action.response.id)

      if (index < 0) return { ...state }

      return {
        ...state,
        status: STATUS_UPDATE_DATA_SUCCESS,
        items: [
          ...state.items.slice(0, index),
          action.response,
          ...state.items.slice(index + 1)
        ]
      }
    }

    case ACTION_DELETE_RESOURCE: {
      const index = state.items.findIndex((p) => p.id === action.id)

      if (index < 0) return { ...state }

      return {
        ...state,
        status: STATUS_UPDATE_DATA_SUCCESS,
        items: [
          ...state.items.slice(0, index),
          ...state.items.slice(index + 1)
        ],
        pager: {
          ...state.pager,
          count: state.pager.count - 1
        }
      }
    }
    default:
      return { ...state }
  }
}

export const createResourceManager = (options = {}) => {
  const ResourcesStateContext = React.createContext()
  const ResourcesDispatchContext = React.createContext()

  const ResourcesProvider = ({ children, value }) => {
    const [state, dispatch] = useReducer(reducer, value)

    return (
      <ResourcesStateContext.Provider value={state}>
        <ResourcesDispatchContext.Provider value={dispatch}>
          {children}
        </ResourcesDispatchContext.Provider>
      </ResourcesStateContext.Provider>
    )
  }

  ResourcesProvider.propTypes = {
    children: PropTypes.any,
    value: PropTypes.any
  }

  const useResourcesState = () => {
    const context = useContext(ResourcesStateContext)

    if (context === undefined) {
      throw new Error(
        'useResourcesState must be used within a ResourcesProvider'
      )
    }

    return context
  }

  const useResourcesDispatch = () => {
    const context = useContext(ResourcesDispatchContext)

    if (context === undefined) {
      throw new Error(
        'useResourcesDispatch must be used within a ResourcesProvider'
      )
    }

    return context
  }

  const getAction = (name) => {
    const actionsOptions = options.actions || {}
    const action = actionsOptions[name]

    if (!action || !action.action) {
      throw new Error(
        `ResourceManager: missing options for the action '${name}'`
      )
    }

    return action
  }

  const onSuccess = (action) => {
    if (action.onSuccess) {
      const message = action.onSuccess()
      if (message) {
        message
      }
    }
  }

  const onError = (action, err) => {
    if (err?.response?.status === 401) return

    if (action.onError) {
      const message = action.onError()
      if (message) {
        message
      }
    }
  }

  const onEnd = (action) => {
    action.onEnd && action.onEnd()
  }

  const useResourcesReducer = () => {
    const state = useResourcesState()
    const dispatch = useResourcesDispatch()

    const loadResources = useCallback(
      async (page = 1) => {
        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_LOADING })

        const action = getAction('loadResources')

        try {
          let response = await action.action(page)

          if (action.preprocess) {
            response = action.preprocess(response)
          }

          dispatch({ type: ACTION_LOAD_RESOURCES, response })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })

          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const loadResource = useCallback(
      async (id) => {
        const action = getAction('loadResource')

        try {
          let response = await action.action(id)

          if (action.preprocess) {
            response = action.preprocess(response)
          }

          dispatch({ type: ACTION_LOAD_RESOURCE, response })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })

          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const createResource = useCallback(
      async (data) => {
        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

        const action = getAction('createResource')

        try {
          let response = await action.action(data)

          if (action.preprocess) {
            response = action.preprocess(response)
          }

          dispatch({ type: ACTION_ADD_RESOURCE, response })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })

          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const editResource = useCallback(
      async (id, data) => {
        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

        const action = getAction('editResource')

        try {
          let response = await action.action(id, data)

          if (action.preprocess) {
            response = action.preprocess(response)
          }

          dispatch({ type: ACTION_EDIT_RESOURCE, response })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })

          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const deleteResource = useCallback(
      async (id) => {
        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

        const action = getAction('deleteResource')

        try {
          await action.action(id)
          dispatch({ type: ACTION_DELETE_RESOURCE, id })

          onSuccess(action)
        } catch (err) {
          dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })
          onError(action, err)
        } finally {
          onEnd(action)
        }
      },
      [dispatch]
    )

    const test = () => {
      dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_ERROR })
    }

    return [
      state,
      {
        loadResources,
        loadResource,
        createResource,
        editResource,
        deleteResource,
        test
      }
    ]
  }

  return {
    INITIAL_STATE: { ...INITIAL_STATE },
    Provider: ResourcesProvider,
    useReducer: useResourcesReducer
  }
}
