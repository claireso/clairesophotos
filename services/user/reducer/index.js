import React, { useReducer, useContext, useCallback } from 'react'
import PropTypes from 'prop-types'
import Router from 'next/router'

import { loginUser, getMe, logout } from '@services/api'

const STATUS_IDLE = 'idle'
const STATUS_PENDING = 'pending'

const STATUS_AUTH_SUCCESS = 'auth_success'
const STATUS_AUTH_ERROR = 'auth_error'

const ACTION_UPDATE_STATUS = 'user/update_status'
const ACTION_SET_USER = 'user/set'

export const INITIAL_STATE = {
  status: STATUS_IDLE
}

const UserStateContext = React.createContext()
const UserDispatchContext = React.createContext()

function reducer(state, action) {
  switch (action.type) {
    case ACTION_UPDATE_STATUS:
      return {
        ...state,
        status: action.status
      }
    case ACTION_SET_USER: {
      const user = action.user

      return {
        ...state,
        status: user ? STATUS_AUTH_SUCCESS : STATUS_IDLE,
        user: user
      }
    }
    default:
      return { ...state }
  }
}

export const UserProvider = ({ children, value }) => {
  const [state, dispatch] = useReducer(reducer, value)

  return (
    <UserStateContext.Provider value={state}>
      <UserDispatchContext.Provider value={dispatch}>
        {children}
      </UserDispatchContext.Provider>
    </UserStateContext.Provider>
  )
}

UserProvider.propTypes = {
  children: PropTypes.any,
  value: PropTypes.object
}

const useUserState = () => {
  const context = useContext(UserStateContext)

  if (context === undefined) {
    throw new Error('useUserState must be used within a UserProvider')
  }

  return context
}

const useUserDispatch = () => {
  const context = useContext(UserDispatchContext)

  if (context === undefined) {
    throw new Error('useUserDispatch must be used within a UserProvider')
  }

  return context
}

export const useUserReducer = () => {
  const state = useUserState()
  const dispatch = useUserDispatch()

  const actionLogin = useCallback(
    async (data) => {
      dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

      try {
        await loginUser(data).ready
        const user = await getMe().ready

        dispatch({ type: ACTION_SET_USER, user })

        Router.push('/admin/photos')
      } catch (err) {
        const status = err?.response?.status

        dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_AUTH_ERROR })

        if ([401, 422].includes(status)) {
          alert('Bad username/password. Please retry')
        }
      }
    },
    [dispatch]
  )

  const actionLogout = useCallback(async () => {
    try {
      await logout().ready
      dispatch({ type: ACTION_SET_USER, undefined })
    } finally {
      Router.push('/admin/login')
    }
  }, [dispatch])

  const actionGetMe = useCallback(async () => {
    dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_PENDING })

    try {
      const user = await getMe().ready

      dispatch({ type: ACTION_SET_USER, user })
    } catch {
      dispatch({ type: ACTION_UPDATE_STATUS, status: STATUS_AUTH_ERROR })
    }
  }, [dispatch])

  return [
    state,
    { login: actionLogin, logout: actionLogout, getMe: actionGetMe }
  ]
}
