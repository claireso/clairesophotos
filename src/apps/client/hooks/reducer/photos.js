export const TYPES = {
  SET_LOADING: 'SET_LOADING',
  GET_PHOTOS_SUCCESS: 'GET_PHOTOS_SUCCESS'
}

export default (state, action) => {
  switch (action.type) {
    case TYPES.SET_LOADING:
      return { ...state, isLoading: action.isLoading }

    case TYPES.GET_PHOTOS_SUCCESS:
      return { ...state, ...action.response }

    default:
      return { ...state }
  }
}
