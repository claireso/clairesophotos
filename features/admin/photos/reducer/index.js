import Router from 'next/router'

import { createResourceManager } from '@services/resources/reducer'
import {
  displaySuccessMessage,
  displayErrorMessage
} from '@services/messages/reducer'
import {
  getPhotos,
  createPhoto,
  deletePhoto,
  editPhoto,
  getPhoto
} from '@services/api'

const PhotosResourceManager = createResourceManager({
  actions: {
    loadResources: {
      action: getPhotos
    },
    loadResource: {
      action: getPhoto
    },
    createResource: {
      action: createPhoto,
      onSuccess: () => {
        displaySuccessMessage({
          message: 'Your photo has been created successfully',
          key: 'CRUD_PHOTO'
        })
        navigateToList({ root: true })
      },
      onError: () => {
        displayErrorMessage({
          message: 'An error has occured during the creation. Please retry',
          key: 'CRUD_PHOTO'
        })
        navigateToList()
      }
    },
    editResource: {
      action: editPhoto,
      onSuccess: () => {
        displaySuccessMessage({
          message: 'Your photo has been updated successfully',
          key: 'CRUD_PHOTO'
        })
        navigateToList()
      },
      onError: () => {
        displayErrorMessage({
          message: 'An error has occured during the update. Please retry',
          key: 'CRUD_PHOTO'
        })
        navigateToList()
      }
    },
    deleteResource: {
      action: deletePhoto,
      onSuccess: () => {
        displaySuccessMessage({
          message: 'Your photo has been deleted successfully',
          key: 'CRUD_PHOTO'
        })
        navigateToList({ root: true })
      },
      onError: () => {
        displayErrorMessage({
          message: 'An error has occured during the deletion. Please retry',
          key: 'CRUD_PHOTO'
        })
        navigateToList()
      }
    }
  }
})

function navigateToList({ root = false } = {}) {
  const _query = {}

  if (!root && Router.router.query.page) {
    _query['page'] = Router.router.query.page
  }

  Router.push({ pathname: '/admin/photos', query: _query })

  window.scrollTo(0, 0)
}

export const ACTION_TYPES = {
  CREATE: 'create_photo',
  EDIT: 'edit_photo',
  DELETE: 'delete_photo'
}

export const {
  Provider: PhotosProvider,
  useReducer: usePhotosReducer,
  INITIAL_STATE
} = PhotosResourceManager
