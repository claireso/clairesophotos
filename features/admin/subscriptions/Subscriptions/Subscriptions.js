import { Fragment, useEffect, useCallback } from 'react'
import { useRouter } from 'next/router'

import { useSubscriptionsReducer, ACTION_TYPES } from '../reducer'

import Loader from '@components/Loader'
import { List, ListHeader } from '@components/admin/List'
import { Heading1 } from '@components/admin/Headings'
import Pager from '@components/admin/Pager'

import Subscription from './components/Subscription'

const Subscriptions = () => {
  const [
    { items: subscriptions, pager, ...state },
    { loadResources }
  ] = useSubscriptionsReducer()
  const isLoading = ['idle', 'loading'].includes(state.status)

  const router = useRouter()
  const { query, pathname } = router

  useEffect(() => {
    loadResources(query.page)
  }, [query.page, loadResources])

  const navigate = useCallback(
    (params) => {
      router.push({ pathname: pathname, query: { ...query, ...params } })
    },
    [pathname, query, router]
  )

  const onPageChange = useCallback(({ page }) => navigate({ page }), [navigate])

  const onDelete = useCallback(
    (id) => {
      navigate({
        action: ACTION_TYPES.DELETE,
        id: id
      })
    },
    [navigate]
  )

  return (
    <Fragment>
      <ListHeader>
        <Heading1>
          Your subscriptions {pager && <span>({pager.count})</span>}
        </Heading1>
      </ListHeader>

      {isLoading ? (
        <Loader />
      ) : (
        <Fragment>
          <List>
            {subscriptions.map((subscription, index) => (
              <Subscription key={index} {...subscription} onDelete={onDelete} />
            ))}
          </List>
          <Pager {...pager} navigate={onPageChange} />
        </Fragment>
      )}
    </Fragment>
  )
}

export default Subscriptions
