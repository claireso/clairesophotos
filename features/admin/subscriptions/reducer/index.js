import Router from 'next/router'

import { createResourceManager } from '@services/resources/reducer'
import {
  displaySuccessMessage,
  displayErrorMessage
} from '@services/messages/reducer'
import { getSubscriptions, deleteSubscription } from '@services/api'

const SubscriptionsResourceManager = createResourceManager({
  actions: {
    loadResources: {
      action: getSubscriptions
    },
    deleteResource: {
      action: deleteSubscription,
      onSuccess: () => {
        displaySuccessMessage({
          message: 'Your subscription has been deleted successfully',
          key: 'CRUD_SUBSCRIPTION'
        })
        navigateToList({ root: true })
      },
      onError: () => {
        displayErrorMessage({
          message: 'An error has occured during the deletion. Please retry',
          key: 'CRUD_SUBSCRIPTION'
        })
        navigateToList()
      }
    }
  }
})

function navigateToList({ root = false } = {}) {
  const _query = {}

  if (!root && Router.router.query.page) {
    _query['page'] = Router.router.query.page
  }

  Router.push({ pathname: '/admin/subscriptions', query: _query })

  window.scrollTo(0, 0)
}

export const ACTION_TYPES = {
  DELETE: 'delete_subscription'
}

export const {
  Provider: SubscriptionsProvider,
  useReducer: useSubscriptionsReducer,
  INITIAL_STATE
} = SubscriptionsResourceManager
