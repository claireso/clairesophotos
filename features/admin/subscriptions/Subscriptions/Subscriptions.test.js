import { render, waitFor, fireEvent } from '@testing-library/react'

import withTestRouter from '@utils/hoc/withTestRouter'

import { SubscriptionsProvider, INITIAL_STATE } from '../reducer'
import Subscriptions from './Subscriptions'

describe('List Subscriptions', () => {
  const renderComponent = (state = INITIAL_STATE, routerProps = {}) => {
    return render(
      withTestRouter(
        <SubscriptionsProvider value={state}>
          <Subscriptions />
        </SubscriptionsProvider>,
        {
          pathname: '/admin/subscriptions',
          ...routerProps
        }
      )
    )
  }

  beforeEach(() => {
    fetch.resetMocks()
    fetch.mockResponseOnce(JSON.stringify(global.__SUBSCRIPTIONS__))
  })

  test('should render loading view [status - idle]', () => {
    const { container } = renderComponent()

    expect(container).toMatchSnapshot()
  })

  test('should render loading view [status - loading]', () => {
    const state = {
      ...INITIAL_STATE,
      status: 'loading'
    }

    const { container } = renderComponent(state)

    expect(container).toMatchSnapshot()
  })

  test('should load and render subscriptions', async () => {
    const { getByText } = renderComponent()

    await waitFor(() => {
      expect(getByText('https://fcm.googleapis.com/')).toBeInTheDocument()
    })
  })

  test('should delete', async () => {
    const push = jest.fn()

    const { getByText, getByTitle } = renderComponent(undefined, { push })

    await waitFor(() => {
      expect(getByText('https://fcm.googleapis.com/')).toBeInTheDocument()
    })

    fireEvent.click(getByTitle('Revoke'))

    expect(push).toHaveBeenCalledWith({
      pathname: '/admin/subscriptions',
      query: {
        action: 'delete_subscription',
        id: 118
      }
    })
  })

  test('should navigate from pager', async () => {
    const push = jest.fn()

    const { getByTitle, getByText } = renderComponent(undefined, { push })

    await waitFor(() => {
      expect(getByText('https://fcm.googleapis.com/')).toBeInTheDocument()
    })

    fireEvent.click(getByTitle(/Next page/i))

    expect(push).toHaveBeenCalledWith({
      pathname: '/admin/subscriptions',
      query: {
        page: 2
      }
    })
  })
})
