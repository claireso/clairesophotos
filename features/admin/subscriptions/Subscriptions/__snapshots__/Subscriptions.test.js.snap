// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Subscriptions should load and render subscriptions 1`] = `
.c2 {
  margin: 0;
  padding: 0;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 2rem;
}

.c0 > h1 {
  line-height: 1.835;
  margin: 0;
}

.c0 > h1 > span {
  color: var(--gray-2);
  font-size: 2rem;
  font-weight: normal;
  line-height: 1;
}

.c1 {
  font-size: 2.4rem;
  margin: 0 0 2.4rem;
}

.c10 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 0.2rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: 'Roboto',Arial,sans-serif;
  font-size: 1.2rem;
  line-height: 1.3333333;
  outline: none;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
  background: var(--secondary);
  color: var(--text);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.4rem;
  height: 4.4rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0;
  width: 4.4rem;
}

.c10 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c10 + .c9 {
  margin: 0 0 0 0.8rem;
}

.c10:hover {
  background: var(--secondary-darker);
}

.c10 + .Buttonsstyles__StyledPrimaryButton-sc-4b7ll8-1 {
  margin-left: 0.8rem;
}

.c7 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 0.2rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: 'Roboto',Arial,sans-serif;
  font-size: 1.2rem;
  line-height: 1.3333333;
  outline: none;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
  background: none;
  color: inherit;
  padding: 0.8rem;
}

.c7 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c7 + .c6 {
  margin: 0 0 0 0.8rem;
}

.c7 > svg {
  margin: 0;
}

.c7 + .c6 {
  margin: 0;
}

.c8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style-type: none;
  width: 100%;
  padding: 0;
}

.c8 > li {
  margin: 0 0.4rem;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1.4rem;
  list-style: none;
  padding: 1.5rem;
  -webkit-transition: background 100ms ease-out;
  transition: background 100ms ease-out;
}

.c4:hover {
  background: var(--gray-5);
}

.c4 + .c3 {
  border-top: 1px solid var(--secondary);
}

.c4 dl {
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  margin: 0;
  padding-right: 4rem;
}

.c4 dt {
  font-weight: 700;
  margin: 0 0 0.5rem;
}

.c4 dd {
  margin: 0 0 1rem;
  word-wrap: break-word;
  word-break: break-all;
}

.c5 {
  -webkit-transition: opacity 150ms ease-out;
  transition: opacity 150ms ease-out;
}

.c5 button {
  opacity: 0.5;
  -webkit-transition: opacity 150ms ease-out;
  transition: opacity 150ms ease-out;
}

.c5 button:hover {
  opacity: 1;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Your subscriptions 
      <span>
        (
        184
        )
      </span>
    </h1>
  </div>
  <ul
    class="c2"
  >
    <li
      class="c3 c4"
    >
      <dl>
        <dt>
          Created at:
        </dt>
        <dd>
          2019-02-05 16:41:23
        </dd>
        <dt>
          Endpoint:
        </dt>
        <dd>
          https://fcm.googleapis.com/
        </dd>
      </dl>
      <p
        class="c5"
      >
        <button
          class="Buttonsstyles__Button-sc-4b7ll8-0 c6 c7"
          title="Revoke"
        >
          <svg
            height="16"
            version="1.1"
            viewBox="0 0 22 28"
            width="22"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 21.5v-11c0-0.281-0.219-0.5-0.5-0.5h-1c-0.281 0-0.5 0.219-0.5 0.5v11c0 0.281 0.219 0.5 0.5 0.5h1c0.281 0 0.5-0.219 0.5-0.5zM12 21.5v-11c0-0.281-0.219-0.5-0.5-0.5h-1c-0.281 0-0.5 0.219-0.5 0.5v11c0 0.281 0.219 0.5 0.5 0.5h1c0.281 0 0.5-0.219 0.5-0.5zM16 21.5v-11c0-0.281-0.219-0.5-0.5-0.5h-1c-0.281 0-0.5 0.219-0.5 0.5v11c0 0.281 0.219 0.5 0.5 0.5h1c0.281 0 0.5-0.219 0.5-0.5zM7.5 6h7l-0.75-1.828c-0.047-0.063-0.187-0.156-0.266-0.172h-4.953c-0.094 0.016-0.219 0.109-0.266 0.172zM22 6.5v1c0 0.281-0.219 0.5-0.5 0.5h-1.5v14.812c0 1.719-1.125 3.187-2.5 3.187h-13c-1.375 0-2.5-1.406-2.5-3.125v-14.875h-1.5c-0.281 0-0.5-0.219-0.5-0.5v-1c0-0.281 0.219-0.5 0.5-0.5h4.828l1.094-2.609c0.313-0.766 1.25-1.391 2.078-1.391h5c0.828 0 1.766 0.625 2.078 1.391l1.094 2.609h4.828c0.281 0 0.5 0.219 0.5 0.5z"
            />
          </svg>
        </button>
      </p>
    </li>
  </ul>
  <ul
    class="c8"
  >
    <li>
      <button
        class="Buttonsstyles__Button-sc-4b7ll8-0 Buttonsstyles__SecondaryButton-sc-4b7ll8-3 c9 c10"
        title="Next page"
      >
        »
      </button>
    </li>
    <li>
      <button
        class="Buttonsstyles__Button-sc-4b7ll8-0 Buttonsstyles__SecondaryButton-sc-4b7ll8-3 c9 c10"
        title="Last page"
      >
        »»
      </button>
    </li>
  </ul>
</div>
`;

exports[`List Subscriptions should render loading view [status - idle] 1`] = `
.c2 {
  margin: 2rem 0;
  -webkit-animation: jBcSpD 800ms ease-in-out infinite alternate;
  animation: jBcSpD 800ms ease-in-out infinite alternate;
}

.c2:after {
  background: var(--primary);
  border-radius: 50%;
  content: '';
  display: block;
  height: 1rem;
  width: 1rem;
  margin: 0 auto;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 2rem;
}

.c0 > h1 {
  line-height: 1.835;
  margin: 0;
}

.c0 > h1 > span {
  color: var(--gray-2);
  font-size: 2rem;
  font-weight: normal;
  line-height: 1;
}

.c1 {
  font-size: 2.4rem;
  margin: 0 0 2.4rem;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Your subscriptions 
      <span>
        (
        )
      </span>
    </h1>
  </div>
  <div
    class="c2"
  />
</div>
`;

exports[`List Subscriptions should render loading view [status - loading] 1`] = `
.c2 {
  margin: 2rem 0;
  -webkit-animation: jBcSpD 800ms ease-in-out infinite alternate;
  animation: jBcSpD 800ms ease-in-out infinite alternate;
}

.c2:after {
  background: var(--primary);
  border-radius: 50%;
  content: '';
  display: block;
  height: 1rem;
  width: 1rem;
  margin: 0 auto;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 2rem;
}

.c0 > h1 {
  line-height: 1.835;
  margin: 0;
}

.c0 > h1 > span {
  color: var(--gray-2);
  font-size: 2rem;
  font-weight: normal;
  line-height: 1;
}

.c1 {
  font-size: 2.4rem;
  margin: 0 0 2.4rem;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Your subscriptions 
      <span>
        (
        )
      </span>
    </h1>
  </div>
  <div
    class="c2"
  />
</div>
`;
